File: Data Structures & Algorithms/search-2d-matrix/submission-1.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        // Determine which row target may be in
        int searchRowIdx = -1;
        for (int i = 0; i < matrix.size(); i++) {
            if (target >= matrix[i][0]) {
                searchRowIdx = i;
            }
        }

        if (searchRowIdx == -1) return false;

        // Perform binary search on row for target
        int L = 0, R = matrix[searchRowIdx].size() - 1;

        while (L <= R) {
            int mid = L + (R - L) / 2;

            if (target > matrix[searchRowIdx][mid]) {
                L = mid + 1;
            } else if (target < matrix[searchRowIdx][mid]) {
                R = mid - 1;
            } else {
                return true;
            }
        }

        return false;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        L, R = 0, len(nums) - 1

        while L <= R:
            mid = L + (R - L) // 2

            if nums[mid] == target:
                return mid

            # Left side is sorted (from L to mid)
            if nums[L] <= nums[mid]:
                if nums[L] <= target < nums[mid]:
                    R = mid - 1
                else:
                    L = mid + 1
            # Right side is sorted (from mid to R)
            else:
                if nums[mid] < target <= nums[R]:
                    L = mid + 1
                else:
                    R = mid - 1

        return -1